sale_mrp_bom/models/sale_order_line.py

```diff
@@ -0,0 +1,34 @@
+# Copyright 2020 Akretion Renato Lima <renato.lima@akretion.com.br>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    bom_id = fields.Many2one(
+        comodel_name="mrp.bom",
+        string="BoM",
+        domain="[('product_tmpl_id.product_variant_ids', '=', product_id),"
+        "'|', ('product_id', '=', product_id), "
+        "('product_id', '=', False)]",
+    )
+
+    @api.constrains("bom_id", "product_id")
+    def _check_match_product_variant_ids(self):
+        for line in self:
+            if line.bom_id:
+                bom_product_tmpl = line.bom_id.product_tmpl_id
+                bom_product = bom_product_tmpl.product_variant_ids
+            else:
+                bom_product_tmpl, bom_product = None, None
+            line_product = line.product_id
+            if not bom_product or line_product == bom_product:
+                continue
+            raise ValidationError(
+                _(
+                    "Please select BoM that has matched product with the line `{}`"
```

Comment:
It's better to use translate function args:

```suggestion
                    "Please select BoM that has matched product with the line `%(product_name)s`", product_name=line_product.name)
```